src/Bicep.Core/Registry/AzureContainerRegistryManager.cs

```diff
@@ -27,104 +35,297 @@ public class AzureContainerRegistryManager
         private const StringComparison DigestComparison = StringComparison.Ordinal;
 
         private readonly IContainerRegistryClientFactory clientFactory;
+        private readonly IFeatureProvider features;
 
-        public AzureContainerRegistryManager(IContainerRegistryClientFactory clientFactory)
+        // https://docs.docker.com/registry/spec/api/#content-digests
+        // "While the algorithm does allow one to implement a wide variety of algorithms, compliant implementations should use sha256."
+        private readonly string DigestAlgorithmIdentifier = DescriptorFactory.AlgorithmIdentifierSha256;
+
+        public AzureContainerRegistryManager(IContainerRegistryClientFactory clientFactory, IFeatureProvider features)
         {
             this.clientFactory = clientFactory;
+            this.features = features;
         }
 
-        public async Task<OciArtifactResult> PullArtifactAsync(RootConfiguration configuration, OciArtifactModuleReference moduleReference)
+        public async Task<OciArtifactResult> PullModuleArtifactsAsync(RootConfiguration configuration, OciArtifactModuleReference moduleReference, bool downloadSource = true)
         {
-            ContainerRegistryContentClient client;
+            IOciRegistryContentClient client;
             OciManifest manifest;
             Stream manifestStream;
-            string manifestDigest;
+            string moduleManifestDigest;
 
-            async Task<(ContainerRegistryContentClient, OciManifest, Stream, string)> DownloadManifestInternalAsync(bool anonymousAccess)
+            async Task<(IOciRegistryContentClient, OciManifest, Stream, string)> DownloadModuleManifestInternalAsync(bool anonymousAccess)
             {
                 var client = this.CreateBlobClient(configuration, moduleReference, anonymousAccess);
-                var (manifest, manifestStream, manifestDigest) = await DownloadManifestAsync(moduleReference, client);
+                var (manifest, manifestStream, manifestDigest) = await DownloadModuleManifestAsync(moduleReference, client);
                 return (client, manifest, manifestStream, manifestDigest);
             }
 
             try
             {
                 // Try authenticated client first.
                 Trace.WriteLine($"Authenticated attempt to pull artifact for module {moduleReference.FullyQualifiedReference}.");
-                (client, manifest, manifestStream, manifestDigest) = await DownloadManifestInternalAsync(anonymousAccess: false);
+                (client, manifest, manifestStream, moduleManifestDigest) = await DownloadModuleManifestInternalAsync(anonymousAccess: false);
```

Comment:
Can we try to keep the notion of Modules decoupled from this class? We require restoring other OCI artifact types that are not modules, such as providers and moving the Module specific logic to a separate handler allows for this kind of code-reuse

Reply:
As I mentioned yesterday, although the class seems to have been intended to be generic, the current code actually contains quite a few specific assumptions that apply only to Bicep modules.  I considered breaking it into two classes, but hated to make that change in a single PR.  I ended up breaking it into two sets of methods, ones that are generic, and ones that aren't (like DownloadManifestInternalAsync) and that *are explicit about their specificity in their names*.

The logical next step would be a separate PR to break it into two classes based on the refactoring I already did.

Reply:
As discussed, I will check in current changes soon, you will then split the class